crates/shared/src/utils.rs

```diff
@@ -148,37 +153,189 @@ impl<Types: NodeType, V: StaticVersionType + 'static> Stream for EventServiceStr
     ) -> Poll<Option<Self::Item>> {
```

Comment:
You can tell me if I'm being overly paranoid here. But I've been bitten by this sort of thing in the past and I find manual future/stream implementations really difficult to understand and debug. Can we not rewrite this whole thing with futures combinators, sort of like

```rust
let stream = Self::connect_inner(self.api_url.clone()).await?;
futures::stream::unfold((self.api_url.clone(), stream), |(api_url, mut stream)| async move {
  loop {
    match stream.next().await {
      Some(Ok(x)) => break Some((x, (api_url, stream))),
      Some(Err(err)) => {
        tracing::warn!(...);
      }
      None => {
        match Self::connect_inner(api_url.clone()).await {
          Ok(new_stream) => stream = new_stream,
          Err(err) => {
            tracing::error!("error connecting, giving up");
            break None;
          }
      }
  }
});
```

I don't know if that's easier or harder for you to reason about. To me it's easier, but I get it's kind of a personal mindset thing. But at least with `unfold` and other combinators you don't have to worry about low-level stuff like forgetting to wake a future or schedule a wakeup.

Reply:
Yeah, I think you're right. This _is_ harder for me to understand at a glance, but has less footgun potential. I'll rewrite it with combinators